### ISSUE TITLE
conda-build doesn't add built package to testing environment

### ISSUE THREAD

**Issue by dschreij** [conda/conda-build#1318]
With versions newer than conda_build 1.20.3 + conda 4.0.9 + conda_env 2.4.5 I have a lot of trouble building packages, which didn't occur with these or prior versions.

For example, I'm trying to build sdl2 and have the weird behavior that the built package is not added to its own testing environment. The build process completes without error, and when I check the build directory (in my case `~/anaconda/conda-bld/sdl2_<some_stamp>/_b_env_placeholder_100x_etc`) I see it has the following contents:

~~~
.
├── bin
│   ├── activate -> /Users/daniel/anaconda/bin/activate
│   ├── conda -> /Users/daniel/anaconda/bin/conda
│   ├── deactivate -> /Users/daniel/anaconda/bin/deactivate
│   └── sdl2-config
├── include
│   └── SDL2
│       ├── SDL.h
│       ├── SDL_assert.h
│       ├── SDL_atomic.h
│       ├── SDL_audio.h
│       ├── SDL_bits.h
│       ├── SDL_blendmode.h
│       ├── SDL_clipboard.h
│       ├── SDL_config.h
│       ├── SDL_cpuinfo.h
│       ├── SDL_egl.h
│       ├── SDL_endian.h
│       ├── SDL_error.h
│       ├── SDL_events.h
│       ├── SDL_filesystem.h
│       ├── SDL_gamecontroller.h
│       ├── SDL_gesture.h
│       ├── SDL_haptic.h
│       ├── SDL_hints.h
│       ├── SDL_joystick.h
│       ├── SDL_keyboard.h
│       ├── SDL_keycode.h
│       ├── SDL_loadso.h
│       ├── SDL_log.h
│       ├── SDL_main.h
│       ├── SDL_messagebox.h
│       ├── SDL_mouse.h
│       ├── SDL_mutex.h
│       ├── SDL_name.h
│       ├── SDL_opengl.h
│       ├── SDL_opengl_glext.h
│       ├── SDL_opengles.h
│       ├── SDL_opengles2.h
│       ├── SDL_opengles2_gl2.h
│       ├── SDL_opengles2_gl2ext.h
│       ├── SDL_opengles2_gl2platform.h
│       ├── SDL_opengles2_khrplatform.h
│       ├── SDL_pixels.h
│       ├── SDL_platform.h
│       ├── SDL_power.h
│       ├── SDL_quit.h
│       ├── SDL_rect.h
│       ├── SDL_render.h
│       ├── SDL_revision.h
│       ├── SDL_rwops.h
│       ├── SDL_scancode.h
│       ├── SDL_shape.h
│       ├── SDL_stdinc.h
│       ├── SDL_surface.h
│       ├── SDL_system.h
│       ├── SDL_syswm.h
│       ├── SDL_test.h
│       ├── SDL_test_assert.h
│       ├── SDL_test_common.h
│       ├── SDL_test_compare.h
│       ├── SDL_test_crc32.h
│       ├── SDL_test_font.h
│       ├── SDL_test_fuzzer.h
│       ├── SDL_test_harness.h
│       ├── SDL_test_images.h
│       ├── SDL_test_log.h
│       ├── SDL_test_md5.h
│       ├── SDL_test_random.h
│       ├── SDL_thread.h
│       ├── SDL_timer.h
│       ├── SDL_touch.h
│       ├── SDL_types.h
│       ├── SDL_version.h
│       ├── SDL_video.h
│       ├── begin_code.h
│       └── close_code.h
├── info
│   ├── about.json
│   ├── files
│   ├── has_prefix
│   ├── index.json
│   └── recipe
│       ├── bld.bat
│       ├── build.sh
│       ├── meta.yaml
│       ├── meta.yaml.template
│       └── run_test.sh
├── lib
│   ├── cmake
│   │   └── SDL2
│   │       └── sdl2-config.cmake
│   ├── libSDL2-2.0.0.dylib
│   ├── libSDL2.a
│   ├── libSDL2.dylib -> libSDL2-2.0.0.dylib
│   ├── libSDL2.la
│   ├── libSDL2_test.a
│   ├── libSDL2main.a
│   └── pkgconfig
│       └── sdl2.pc
└── share
    └── aclocal
        └── sdl2.m4

11 directories, 92 files
~~~

However, the test, which simply calls `bin/sdl2-config --version` fails because the sdl2-config command is not found in the testing environment. With the latest versions (conda 4.1.11-py27_1 + conda_env 2.5.2-py27_0+ conda_build 2.0.0-py27_0) I get to see:

~~~
TEST START: sdl2-2.0.4-0
updating index in: /Users/daniel/anaconda/conda-bld/osx-64
updating index in: /Users/daniel/anaconda/conda-bld/noarch

The following NEW packages will be INSTALLED:

    sdl2: 2.0.4-0 local

+ source /Users/daniel/anaconda/bin/activate /Users/daniel/anaconda/conda-bld/sdl2_1472985972179/_t_env
+ /bin/bash -x -e /Users/daniel/anaconda/conda-bld/sdl2_1472985972179/test_tmp/run_test.sh
+ /Users/daniel/anaconda/conda-bld/sdl2_1472985972179/_t_env/bin/sdl2-config --version
/Users/daniel/anaconda/conda-bld/sdl2_1472985972179/test_tmp/run_test.sh: line 3: /Users/daniel/anaconda/conda-bld/sdl2_1472985972179/_t_env/bin/sdl2-config: No such file or directory
TESTS FAILED: sdl2-2.0.4-0
~~~

There is a mention that the just built sdl2 is being installed, but this does not appear to happen. I know that I'm using versions from conda-forge, but this also happens with the latest versions from the main default conda channel. If I check the contents of the `~/anaconda/conda-bld/sdl2_<some_stamp>/_t_env` folder, it seems to be a default, empty folder. It only contains

~~~
.
├── bin
│   ├── activate -> /Users/daniel/anaconda/bin/activate
│   ├── conda -> /Users/daniel/anaconda/bin/conda
│   └── deactivate -> /Users/daniel/anaconda/bin/deactivate
└── conda-meta
    ├── history
    └── sdl2-2.0.4-0.json
~~~

Shouldn't this environment be similar to the build environment?

I do not have these problems with the aforementioned and older versions of conda, conda-build and conda-env. A discrepancy can also be seen in the console output when the testing phase begins: With the older conda versions, the sdl2 package is explicitly installed in the test environment:

~~~
TEST START: sdl2-2.0.4-0
Using Anaconda API: https://api.anaconda.org
Fetching package metadata: ..........

WARNING: conda-build appears to be out of date. You have version 1.20.3 but the
latest version is 2.0.0. Run

conda update -n root conda-build

to get the latest version.

.Solving package specifications: .........

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    sdl2-2.0.4                 |                0         1.0 MB

The following NEW packages will be INSTALLED:

    sdl2: 2.0.4-0

Fetching packages ...
sdl2-2.0.4-0.t 100% |############################################################################################################################################################| Time: 0:00:00  63.96 MB/s
Extracting packages ...
[      COMPLETE      ]|###############################################################################################################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|###############################################################################################################################################################################| 100%
+ /Users/daniel/anaconda/envs/_test/bin/sdl2-config --version
2.0.4
TEST END: sdl2-2.0.4-0
~~~

The `~/anaconda/envs/_test` folder indeed appears to contain the same files as the aforementioned build folder:

~~~
.
├── bin
│   ├── activate -> /Users/daniel/anaconda/bin/activate
│   ├── conda -> /Users/daniel/anaconda/bin/conda
│   ├── deactivate -> /Users/daniel/anaconda/bin/deactivate
│   └── sdl2-config
├── conda-meta
│   ├── history
│   └── sdl2-2.0.4-0.json
├── include
│   └── SDL2
│       ├── SDL.h
│       ├── SDL_assert.h
│       ├── SDL_atomic.h
│       ├── SDL_audio.h
│       ├── SDL_bits.h
│       ├── SDL_blendmode.h
│       ├── SDL_clipboard.h
│       ├── SDL_config.h
│       ├── SDL_cpuinfo.h
│       ├── SDL_egl.h
│       ├── SDL_endian.h
│       ├── SDL_error.h
│       ├── SDL_events.h
│       ├── SDL_filesystem.h
│       ├── SDL_gamecontroller.h
│       ├── SDL_gesture.h
│       ├── SDL_haptic.h
│       ├── SDL_hints.h
│       ├── SDL_joystick.h
│       ├── SDL_keyboard.h
│       ├── SDL_keycode.h
│       ├── SDL_loadso.h
│       ├── SDL_log.h
│       ├── SDL_main.h
│       ├── SDL_messagebox.h
│       ├── SDL_mouse.h
│       ├── SDL_mutex.h
│       ├── SDL_name.h
│       ├── SDL_opengl.h
│       ├── SDL_opengl_glext.h
│       ├── SDL_opengles.h
│       ├── SDL_opengles2.h
│       ├── SDL_opengles2_gl2.h
│       ├── SDL_opengles2_gl2ext.h
│       ├── SDL_opengles2_gl2platform.h
│       ├── SDL_opengles2_khrplatform.h
│       ├── SDL_pixels.h
│       ├── SDL_platform.h
│       ├── SDL_power.h
│       ├── SDL_quit.h
│       ├── SDL_rect.h
│       ├── SDL_render.h
│       ├── SDL_revision.h
│       ├── SDL_rwops.h
│       ├── SDL_scancode.h
│       ├── SDL_shape.h
│       ├── SDL_stdinc.h
│       ├── SDL_surface.h
│       ├── SDL_system.h
│       ├── SDL_syswm.h
│       ├── SDL_test.h
│       ├── SDL_test_assert.h
│       ├── SDL_test_common.h
│       ├── SDL_test_compare.h
│       ├── SDL_test_crc32.h
│       ├── SDL_test_font.h
│       ├── SDL_test_fuzzer.h
│       ├── SDL_test_harness.h
│       ├── SDL_test_images.h
│       ├── SDL_test_log.h
│       ├── SDL_test_md5.h
│       ├── SDL_test_random.h
│       ├── SDL_thread.h
│       ├── SDL_timer.h
│       ├── SDL_touch.h
│       ├── SDL_types.h
│       ├── SDL_version.h
│       ├── SDL_video.h
│       ├── begin_code.h
│       └── close_code.h
├── lib
│   ├── cmake
│   │   └── SDL2
│   │       └── sdl2-config.cmake
│   ├── libSDL2-2.0.0.dylib
│   ├── libSDL2.a
│   ├── libSDL2.dylib -> libSDL2-2.0.0.dylib
│   ├── libSDL2.la
│   ├── libSDL2_test.a
│   ├── libSDL2main.a
│   └── pkgconfig
│       └── sdl2.pc
└── share
    └── aclocal
        └── sdl2.m4
~~~

It may be that there has been an API change in conda-build, or in the way the packages that are required in the test environment need to explicitly mentioned in the meta.yaml file, but this has not been clearly documented.

**Comment by msarahan** [comment]
This is a bug, not an API change, unless you are somehow driving conda-build using it internal functions.  I have been working to identify the source, but have not isolated it yet.  What appears to be the problem is that Conda is keeping old, failed builds in the pkgs cache, and never removing them prior to testing.  See #1309 for some work on that.

Side note: the progress bar output is hidden by default in conda-build 2.0.  Though this is sometimes good visual feedback when using conda-build interactively, it also obfuscates CI logs, which break the output into many lines, rather than keeping each progress bar on one.

I will test with conda 4.0.x, as I have not yet explored that dimension.

**Comment by dschreij** [comment]
Good to know, thanks. I am just building a package using meta.yaml and build scripts, and not using any internal functions!

**Comment by msarahan** [comment]
I believe that this is fixed in #1309 - which is part of the 2.0.1 release.  Please update and let me know if the problem persists.  Closing this issue for now, but happy to reopen if it's not fixed for you.
